dxr/plugins/js/analyze_js/analyze_file.js

```diff
@@ -0,0 +1,727 @@
+    case "ObjectPattern":
+      for (let prop of pat.properties) {
+        this.pattern(prop.value);
+      }
+      break;
+
+    case "ArrayPattern":
+      for (let e of pat.elements) {
+        if (e) {
+          this.pattern(e);
+        }
+      }
+      break;
+
+    case "SpreadExpression":
+      this.pattern(pat.expression);
+      break;
+
+    case "AssignmentExpression":
+      this.pattern(pat.left);
+      this.expression(pat.right);
+      break;
+
+    default:
+      console.log(`Unexpected pattern: ${pat.type} ${JSON.stringify(pat)}`);
+      break;
+    }
+  },
+};
+
+// Comment out some mozilla-specific preprocessor headers in js files.
+function preprocess(text, comment)
+{
+  let substitution = false;
+  let lines = text.split("\n");
+  let preprocessedLines = [];
+  let branches = [true];
+  for (let i = 0; i < lines.length; i++) {
+    let line = lines[i];
+    if (substitution) {
+      line = line.replace(/@(\w+)@/, "''");
+    }
+    let tline = line.trim();
+    if (tline.startsWith("#ifdef") || tline.startsWith("#ifndef") || tline.startsWith("#if ")) {
+      preprocessedLines.push(comment(tline));
+      branches.push(branches[branches.length-1]);
+    } else if (tline.startsWith("#else") ||
+               tline.startsWith("#elif") ||
+               tline.startsWith("#elifdef") ||
+               tline.startsWith("#elifndef")) {
+      preprocessedLines.push(comment(tline));
+      branches.pop();
+      branches.push(false);
+    } else if (tline.startsWith("#endif")) {
+      preprocessedLines.push(comment(tline));
+      branches.pop();
+    } else if (!branches[branches.length-1]) {
+      preprocessedLines.push(comment(tline));
+    } else if (tline.startsWith("#include")) {
+      /*
+      let match = tline.match(/#include "?([A-Za-z0-9_.-]+)"?/);
+      if (!match) {
+        throw new Error(`Invalid include directive: ${filename}:${i+1}`);
+      }
+      let incfile = match[1];
+      preprocessedLines.push(`PREPROCESSOR_INCLUDE("${incfile}");`);
+      */
+      preprocessedLines.push(comment(tline));
+    } else if (tline.startsWith("#filter substitution")) {
+      preprocessedLines.push(comment(tline));
+      substitution = true;
+    } else if (tline.startsWith("#filter")) {
+      preprocessedLines.push(comment(tline));
+    } else if (tline.startsWith("#expand")) {
+      preprocessedLines.push(line.substring(String("#expand ").length));
+    } else if (tline.startsWith("#")) {
+      preprocessedLines.push(comment(tline));
+    } else {
+      preprocessedLines.push(line);
+    }
+  }
+
+  return preprocessedLines.join("\n");
+}
+
+function analyzeJS(filepath, relpath, tempFilepath)
+{
+  fileIndex = relpath;
+  nextSymId = 0;
+  outLines = [];
+  let text = preprocess(String(fs.readFileSync(filepath)), line => "//" + line);
+  try {
+    let ast = esprima.parse(text, {loc: true, source: path.basename(filepath), line: 1, sourceType: "script"});
+    if (ast) {
+        Analyzer.program(ast);
```

Comment:
I notice we don't make a `new Analyzer`. Are some of the hashes Analyzer keeps around important for maintaining cross-file relationships?